addons/l10n_hu_edi/models/l10n_hu_edi_transaction.py

```diff
@@ -0,0 +1,531 @@
+
+    def _search_is_active(self, operator, value):
+        if (operator, value) in [('=', True), ('!=', False)]:
+            return [('state', 'in', _ACTIVE_STATES)]
+        elif (operator, value) in [('=', False), ('!=', True)]:
+            return [('state', 'not in', _ACTIVE_STATES)]
+        else:
+            raise UserError(_('Invalid operator!'))
+
+    def _get_action_info(self):
+        return {
+            'upload': {
+                'start_states': ['to_send', 'token_error'],
+                'end_states': ['token_error', 'sent', 'send_error', 'send_timeout'],
+            },
+            'abort': {
+                'start_states': ['to_send', 'token_error'],
+                'end_states': ['unsent'],
+            },
+            'query_status': {
+                'start_states': ['sent', 'query_error'],
+                'end_states': ['sent', 'confirmed', 'confirmed_warning', 'rejected', 'query_error'],
+            },
+            'recover_timeout': {
+                'start_states': ['send_timeout'],
+                'end_states': ['to_send', 'sent', 'confirmed', 'confirmed_warning', 'rejected', 'query_error', 'send_timeout'],
+            },
+        }
+
+    def _can_perform(self, action):
+        """ Returns whether the given action ('upload', 'abort' or 'query_status') may be performed
+        on a given transaction, given its state.
+
+        You MUST call this method to check whether the action can be performed, before calling the action.
+
+        This ensures integrity of the state machine."""
+        if not self:
+            return True
+        self.ensure_one()
+        return self.state in self._get_action_info().get(action)['start_states']
+
+    # === State transitions === #
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """ When creating a transaction for an invoice, generate the XML. """
+        if any(not vals.get(field_name) for field_name in ['invoice_id', 'credentials_id', 'operation', 'attachment_file'] for vals in vals_list):
+            raise UserError(_('New transactions must reference an invoice, NAV credentials, the invoice operation, and the invoice XML to send!'))
+        transactions = super().create(vals_list)
+        # Set name & mimetype on newly-created attachments.
+        attachments = self.env['ir.attachment'].search([
+            ('res_model', '=', self._name),
+            ('res_id', 'in', transactions.ids),
+            ('res_field', '=', 'attachment_file'),
+        ])
+        for attachment in attachments:
+            attachment.write({
+                'name': self.browse(attachment.res_id)._get_xml_file_name(),
+                'mimetype': 'application/xml',
+            })
+        return transactions
+
+    @check_state_machine(action='upload')
+    def upload(self):
+        """ Send the invoice XMLs of the transactions in `self` to NAV. """
+        with self._acquire_lock():
```

Comment:
Lock should be handled by account_move_send

Reply:
I'll check the locking mechanism on account.move.send

Reply:
Seems like account.move.send doesn't do any locking by default (at least I couldn't find it), so I'm keeping this for now.